File: Dante2/Module2.02/Zadanie2.09/main.c
#include <stdio.h>
#include <math.h>


struct complex_t
{
    float re;
    float im;
};

int read(struct complex_t* cp);
struct complex_t* set(struct complex_t* cp, float re, float im, int *err_code);
float complex_abs(struct complex_t* cp, int * err_code);
void show(const struct complex_t* cp);

int add(const struct complex_t *first, const struct complex_t *second, struct complex_t* result);
int subtract(const struct complex_t *first, const struct complex_t *second, struct complex_t* result);
int multiply(const struct complex_t *first, const struct complex_t *second, struct complex_t* result);
int divide(const struct complex_t *first, const struct complex_t *second, struct complex_t* result);

int main()
{
    int error;
    struct complex_t first;
    struct complex_t second;

    printf("Enter: ");
    error = read(&first);
    if(error != 0)
    {
        printf("Incorrect input");
        return 1;
    }

    printf("Enter: ");
    error = read(&second);
    if(error != 0)
    {
        printf("Incorrect input");
        return 1;
    }

    int err_code1;
    float module1 = complex_abs(&first, &err_code1);
    if(module1 < 0)
    {
        printf("Incorrect input");
        return 1;
    }

    struct complex_t result;
    int result_of_operations;

    add(&first, &second, &result);

    show(&result);

    subtract(&first, &second, &result);

    show(&result);

    multiply(&first, &second, &result);

    show(&result);

    result_of_operations = divide(&first, &second, &result);
    if(result_of_operations == 2)
    {
        printf("Operation not permitted");
        return 0;
    }
    /*if(result_of_operations == 1)
    {
        printf("Incorrect input");
        return 1;
    }*/
    else
    {
        show(&result);
    }

    return 0;
}

int add(const struct complex_t *first, const struct complex_t *second, struct complex_t* result)
{
    if(first == NULL || second == NULL || result == NULL)
    {
        return 1;
    }

    result->re = first->re + second->re;
    result->im = first->im + second->im;

    return 0;
}

int subtract(const struct complex_t *first, const struct complex_t *second, struct complex_t* result)
{
    if(first == NULL || second == NULL || result == NULL)
    {
        return 1;
    }

    result->re = first->re - second->re;
    result->im = first->im - second->im;

    return 0;
}

int multiply(const struct complex_t *first, const struct complex_t *second, struct complex_t* result)
{
    if(first == NULL || second == NULL || result == NULL)
    {
        return 1;
    }

    result->re = first->re * second->re - first->im * second->im;
    result->im = first->re * second->im + first->im * second->re;

    return 0;
}

int divide(const struct complex_t *first, const struct complex_t *second, struct complex_t* result)
{
    if (first == NULL || second == NULL || result == NULL)
    {
        return 1;
    }

    float some_shit_idk_help_me = second->re * second->re + second->im * second->im;
    if (some_shit_idk_help_me <= 0)
    {
        return 2;
    }

    result->re = (first->re * second->re + first->im * second->im) / some_shit_idk_help_me;
    result->im = (first->im * second->re - first->re * second->im) / some_shit_idk_help_me;

    return 0;
}

int read(struct complex_t* cp)
{
    if(cp == NULL)
    {
        return 2;
    }

    float re, im;
    char oper;

    int error;

    error = scanf("%f%ci%f",&re,&oper,&im);
    if(error != 3 || (oper != '-' && oper != '+') || im < 0)
    {
        return 1;
    }

    while(getchar() != '\n');

    int error_code;

    if(oper == '-')
    {
        im = -1 * im;
        if(set(cp, re, im, &error_code) == NULL)
        {
            return 1;
        }
    }
    else if(oper == '+')
    {
        if(set(cp, re, im, &error_code) == NULL)
        {
            return 1;
        }
    }
    else
    {
        return 1;
    }

    return 0;
}

float complex_abs(struct complex_t* cp, int *err_code)
{
    if(cp == NULL)
    {
        if(err_code != NULL)
        {
            *err_code = 1;
        }
        return -1.0f;
    }

    float module = sqrtf(cp->re * cp->re + cp->im * cp->im);
    if(err_code == NULL)
    {
        return module;
    }

    *err_code = 0;
    return module;
}

void show(const struct complex_t* cp)
{
    if(cp == NULL)
    {
        return;
    }

    if(cp->im < 0)
    {
        printf("%.2f - i%.2f\n",cp->re,fabsf(cp->im));
    }
    else
    {
        printf("%.2f + i%.2f\n",cp->re,fabsf(cp->im));
    }
}

struct complex_t* set(struct complex_t* cp, float re, float im, int *err_code)
{
    if(cp == NULL)
    {
        if(err_code != NULL)
        {
            *err_code = 1;
        }
        return NULL;
    }

    cp->im = im;
    cp->re = re;

    if(err_code != NULL)
    {
        *err_code = 0;
    }
    return cp;
}

File: Dante2/Module2.02/Zadanie2.09/main.rs
#[derive(Default)]
struct ComplexT {
    re: f32,
    im: f32,
}

fn main() {
    let mut err: i32;

    let mut first: ComplexT = Default::default();
    println!("Enter: ");
    err = read(&mut first);
    if err != 0 {
        panic!("Error");
    }

    let mut second: ComplexT = Default::default();
    println!("Enter: ");
    err = read(&mut second);
    if err != 0 {
        panic!("Error");
    }

    let module: f32 = complex_abs(&first);
    if module < 0.0 {
        panic!("Error");
    }

    let mut result: ComplexT = Default::default();

    add(&first, &second, &mut result);
    show(&result);

    subtract(&first, &second, &mut result);
    show(&result);

    multiply(&first, &second, &mut result);
    show(&result);

    err = divide(&first, &second, &mut result);
    if err == 2 {
        panic!("Error");
    }
    show(&result);
}

fn read(cp: &mut ComplexT) -> i32 {
    let mut input: String = String::new();

    std::io::stdin().read_line(&mut input).expect("Error");

    let input = input.trim();
    let (re, oper, im) = if let Some(pos) = input.find('+') {
        let re_str = &input[..pos];
        let im_str = &input[pos + 1..];
        (re_str, '+', im_str)
    } else if let Some(pos) = input.find('-') {
        let re_str = &input[..pos];
        let im_str = &input[pos + 1..];
        (re_str, '-', im_str)
    } else {
        panic!("Error");
    };

    let re: f32 = match re.parse() {
        Ok(num) => num,
        Err(_) => {
            panic!("Error");
        }
    };

    let mut im: f32 = match im.parse() {
        Ok(num) => num,
        Err(_) => {
            panic!("Error");
        }
    };

    if oper == '-' {
        im = -1.0 * im;
        set(cp, re, im);
    } else if oper == '+' {
        set(cp, re, im);
    }
    else {
        return 1;
    }

    0
}
fn complex_abs(cp: & ComplexT) -> f32 {
    if cp.re == 0.0 && cp.im == 0.0 {
        return -1.0;
    }
    let module: f32 = f32::sqrt(cp.re * cp.re + cp.im * cp.im);
    module
}
fn show(cp: & ComplexT) {
    if cp.im < 0.0 {
        println!("{:.2} - i{:.2}", cp.re, f32::abs(cp.im));
    }
    else {
        println!("{:.2} + i{:.2}", cp.re, f32::abs(cp.im));
    }
}
fn set(cp: &mut ComplexT, re: f32, im: f32) -> &mut ComplexT {
    cp.re = re;
    cp.im = im;

    cp
}

fn add(first: & ComplexT, second: & ComplexT, result: &mut ComplexT) -> i32 {
    result.re = first.re + second.re;
    result.im = result.im + result.im;
    0
}
fn subtract(first: & ComplexT, second: & ComplexT, result: &mut ComplexT) -> i32 {
    result.re = first.re - second.re;
    result.im = result.im - result.im;
    0
}
fn multiply(first: & ComplexT, second: & ComplexT, result: &mut ComplexT) -> i32 {
    result.re = first.re * second.re - first.im * second.im;
    result.im = first.re * second.im + first.im * second.re;
    0
}
fn divide(first: & ComplexT, second: & ComplexT, result: &mut ComplexT) -> i32 {
    let some_shit_idk_help_me: f32 = second.re * second.re + second.im * second.im;
    if some_shit_idk_help_me <= 0.0 {
        return 2;
    }

    result.re = (first.re * second.re + first.im * second.im) / some_shit_idk_help_me;
    result.im = (first.im * second.re - first.re * second.im) / some_shit_idk_help_me;
    0
}
